org.jacoco.core/src/org/jacoco/core/instr/Instrumenter.java

```diff
@@ -216,6 +225,20 @@ private int instrumentZip(final InputStream input,
 			}
 			zipout.closeEntry();
```

Comment:
Perhaps OfflineInstrumentationCompanionAccessGenerator should have an end/close method and a constructor that accepts an IClassFileWriter with `writeClassFile(String name, byte[] bytes)`.  That way, the logic for when to write the companion classes could be encapsulated in OfflineInstrumentationCompanionAccessGenerator.